impeller/renderer/backend/vulkan/blit_command_vk.cc

```diff
@@ -288,55 +292,63 @@ bool BlitGenerateMipmapCommandVK::Encode(CommandEncoderVK& encoder) const {
     return false;
   }
 
-  // Transition the base mip level to transfer-src layout so we can read from
-  // it and transition the rest to dst-optimal since they are going to be
-  // written to.
   InsertImageMemoryBarrier(
       cmd,                                   // command buffer
       image,                                 // image
       vk::AccessFlagBits::eTransferWrite,    // src access mask
       vk::AccessFlagBits::eTransferRead,     // dst access mask
       src.GetLayout(),                       // old layout
-      vk::ImageLayout::eTransferSrcOptimal,  // new layout
-      vk::PipelineStageFlagBits::eTransfer,  // src stage
-      vk::PipelineStageFlagBits::eTransfer,  // dst stage
-      0u                                     // mip level
-  );
-  InsertImageMemoryBarrier(
-      cmd,                                   // command buffer
-      image,                                 // image
-      {},                                    // src access mask
-      vk::AccessFlagBits::eTransferWrite,    // dst access mask
-      vk::ImageLayout::eUndefined,           // old layout
       vk::ImageLayout::eTransferDstOptimal,  // new layout
       vk::PipelineStageFlagBits::eTransfer,  // src stage
       vk::PipelineStageFlagBits::eTransfer,  // dst stage
-      1u,                                    // mip level
-      mip_count - 1                          // mip level count
-  );
+      0u,                                    // mip level
+      mip_count);
 
-  // Blit from the base mip level to all other levels.
+  vk::ImageMemoryBarrier barrier;
+  barrier.image = image;
+  barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+  barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+  barrier.subresourceRange.aspectMask = vk::ImageAspectFlagBits::eColor;
+  barrier.subresourceRange.baseArrayLayer = 0;
+  barrier.subresourceRange.layerCount = 1;
+  barrier.subresourceRange.levelCount = 1;
+
+  // Blit from the mip level N - 1 to mip level N.
+  size_t width = size.width;
+  size_t height = size.height;
   for (size_t mip_level = 1u; mip_level < mip_count; mip_level++) {
-    vk::ImageBlit blit;
+    barrier.subresourceRange.baseMipLevel = mip_level - 1;
+    barrier.oldLayout = vk::ImageLayout::eTransferDstOptimal;
+    barrier.newLayout = vk::ImageLayout::eTransferSrcOptimal;
+    barrier.srcAccessMask = vk::AccessFlagBits::eTransferWrite;
+    barrier.dstAccessMask = vk::AccessFlagBits::eTransferRead;
 
+    cmd.pipelineBarrier(vk::PipelineStageFlagBits::eTransfer,
```

Comment:
Just to make this easier to read, add a comment about why the barrier is necessary. I'll attempt to jot down how I read it but please proofread.

Reply:
Comment Suggestion:
> We just finished writing to the previous (N-1) mip level. These were initialized to `TransferDst` earler. We are now going to read from it to write to the current level (N) . So it must be converted to `TransferSrc`.

Reply:
Done